spec/models/sample_spec.rb

```diff
@@ -0,0 +1,121 @@
+      )
+      # Remove input files from the sample.
+      @sample.input_files.destroy_all()
+      return @sample
+    end
+
+    context "fetches metadata for and downloads files successfully" do
+      before do
+        @sample = create_basespace_sample
+
+        allow_any_instance_of(BasespaceHelper).to receive(:files_for_basespace_dataset).and_return(
+          [
+            {
+              name: "file 1",
+              download_path: "download_path 1",
+              source_path: "source_path 1"
+            },
+            {
+              name: "file 2",
+              download_path: "download_path 2",
+              source_path: "source_path 2"
+            }
+          ]
+        )
+
+        allow_any_instance_of(BasespaceHelper).to receive(:download_from_basespace_to_s3).and_return(true)
+      end
+
+      it "creates the sample as expected" do
+        @sample.initiate_basespace_file_transfer
+
+        expect(@sample.status).to eq(Sample::STATUS_UPLOADED)
+        # Verify that basespace access token has been removed.
+        expect(@sample.basespace_access_token).to be nil
+        expect(@sample.pipeline_runs.length).to be 0
+        expect(@sample.input_files.length).to be 2
+
+        expect(@sample.input_files[0].name).to eq("file 1")
+        expect(@sample.input_files[0].source_type).to eq(InputFile::SOURCE_TYPE_BASESPACE)
+        expect(@sample.input_files[0].source).to eq("source_path 1")
+        expect(@sample.input_files[1].name).to eq("file 2")
+        expect(@sample.input_files[1].source_type).to eq(InputFile::SOURCE_TYPE_BASESPACE)
+        expect(@sample.input_files[1].source).to eq("source_path 2")
+      end
+    end
+
+    context "cannot fetch files for basespace dataset" do
+      before do
+        @sample = create_basespace_sample
+
+        allow_any_instance_of(BasespaceHelper).to receive(:files_for_basespace_dataset).and_return(nil)
+        allow_any_instance_of(BasespaceHelper).to receive(:download_from_basespace_to_s3).and_return(true)
+      end
+
+      it "fails gracefully and adds a failed pipeline run" do
+        @sample.initiate_basespace_file_transfer
+
+        expect(@sample.status).to eq(Sample::STATUS_CHECKED)
+        # Verify that basespace access token has been removed.
+        expect(@sample.basespace_access_token).to be nil
+        expect(@sample.pipeline_runs.length).to be 1
+        expect(@sample.pipeline_runs[0].results_finalized).to eq(PipelineRun::FINALIZED_FAIL)
+        expect(@sample.pipeline_runs[0].error_message).to eq(PipelineRun::ERROR_BASESPACE_UPLOAD_FAILED)
+        expect(@sample.input_files.length).to be 0
+      end
+    end
+
+    context "cannot download basespace files" do
+      before do
+        @sample = create_basespace_sample
+
+        allow_any_instance_of(BasespaceHelper).to receive(:files_for_basespace_dataset).and_return(
+          [
+            {
+              name: "file 1",
+              download_path: "download_path 1",
```

Comment:
Same note about download path here. Since you are repeating this fake value, maybe it would be nice to extract it to a constant to avoid repeating it.